src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/PageActionInvoker.cs

```diff
@@ -159,28 +159,27 @@ private Task BindArgumentsAsync()
                 await CacheEntry.PropertyBinder(_pageContext, _instance);
             }
 
-            if (_handler == null)
+            if (_handler == null || _handler.Parameters.Count == 0)
             {
                 return;
             }
 
-            var valueProvider = await CompositeValueProvider.CreateAsync(_pageContext, _pageContext.ValueProviderFactories);
+            Debug.Assert(
+                _actionDescriptor.HandlerMethods.Count == CacheEntry.HandlerBinders.Length,
+                "Handler has parameters, but binders are not set up correctly.");
 
-            for (var i = 0; i < _handler.Parameters.Count; i++)
+            PageHandlerBinderDelegate handlerBinder = null;
+            for (var i = 0; i < _actionDescriptor.HandlerMethods.Count; i++)
             {
-                var parameter = _handler.Parameters[i];
-
-                var result = await _parameterBinder.BindModelAsync(
```

Comment:
ParameterBinder was a new thing we invented during the pages design process to serve as a facade over model binding. It shouldn't be pushed to the background, that's just adding a layer. 

What are the gaps that need to be addressed?

Reply:
We still call `ParameterBinder`, but doing this inlines the behavior with binding action methods (https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerBinderDelegateProvider.cs). It also avoids yet another lookup per bind due to https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/ParameterBinder.cs#L185-L190.